File: dotnet-examples/scenario2-data-pipeline/linq-optimization/after/DataProcessingService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Extensions.Logging;
using System.Collections.Concurrent;

namespace Enterprise.DataProcessing.Services
{
    /// <summary>
    /// Service for data processing operations using optimized LINQ.
    /// This implementation demonstrates best practices for efficient LINQ usage.
    /// 
    /// GitHub Copilot Agent Autonomous Mode:
    /// To use with Copilot Agent, instruct it to: "Analyze this class for LINQ performance
    /// bottlenecks and suggest improvements for handling large datasets."
    /// </summary>
    public class DataProcessingService
    {
        private readonly ILogger<DataProcessingService> _logger;

        public DataProcessingService(ILogger<DataProcessingService> logger)
        {
            _logger = logger ?? throw new ArgumentNullException(nameof(logger));
        }

        /// <summary>
        /// Process transaction data to calculate total amount by category.
        /// Optimized implementation using GroupBy with aggregation.
        /// </summary>
        public Dictionary<string, decimal> CalculateTotalByCategory(IEnumerable<Transaction> transactions)
        {
            if (transactions == null || !transactions.Any())
            {
                return new Dictionary<string, decimal>();
            }
            
            // Single pass using GroupBy with aggregation
            return transactions
                .Where(t => t.Amount > 0)
                .GroupBy(t => t.Category)
                .ToDictionary(
                    g => g.Key,
                    g => g.Sum(t => t.Amount)
                );
        }
        
        /// <summary>
        /// Find transactions that meet specific criteria.
        /// Optimized implementation with a single LINQ query.
        /// </summary>
        public IReadOnlyList<Transaction> FindTransactionsByCriteria(IEnumerable<Transaction> transactions, 
                                                                   string merchant, 
                                                                   DateTime startDate, 
                                                                   DateTime endDate, 
                                                                   decimal minAmount)
        {
            if (transactions == null || !transactions.Any())
            {
                return Array.Empty<Transaction>();
            }
            
            // Single LINQ query combining all filters
            return transactions
                .Where(t => t.Merchant.Equals(merchant, StringComparison.OrdinalIgnoreCase) &&
                           t.Date >= startDate && t.Date <= endDate &&
                           t.Amount >= minAmount)
                .ToList();
        }
        
        /// <summary>
        /// Group transactions by month and calculate monthly statistics.
        /// Optimized implementation with efficient grouping.
        /// </summary>
        public IReadOnlyList<MonthlyStatistics> CalculateMonthlyStatistics(IEnumerable<Transaction> transactions)
        {
            if (transactions == null || !transactions.Any())
            {
                return Array.Empty<MonthlyStatistics>();
            }
            
            // Group by year and month with a single LINQ query
            return transactions
                .GroupBy(t => new { t.Date.Year, t.Date.Month })
                .Select(g => new MonthlyStatistics
                {
                    Year = g.Key.Year,
                    Month = g.Key.Month,
                    TransactionCount = g.Count(),
                    TotalAmount = g.Sum(t => t.Amount),
                    AverageAmount = g.Average(t => t.Amount),
                    LargestTransactionAmount = g.Max(t => t.Amount)
                })
                .ToList();
        }
        
        /// <summary>
        /// Process a large list of orders using optimized LINQ.
        /// Efficiently processes orders in a single pass where possible.
        /// </summary>
        public IReadOnlyList<OrderSummary> ProcessOrders(IEnumerable<Order> orders)
        {
            if (orders == null || !orders.Any())
            {
                return Array.Empty<OrderSummary>();
            }
            
            return orders.Select(order =>
            {
                // Calculate total price and item count in one pass
                var orderItems = order.Items;
                var (totalPrice, itemCount) = orderItems.Aggregate(
                    (TotalPrice: 0m, ItemCount: 0),
                    (acc, item) => (
                        acc.TotalPrice + (item.Price * item.Quantity),
                        acc.ItemCount + item.Quantity
                    )
                );
                
                // Calculate discount
                var discount = totalPrice > 1000 ? totalPrice * 0.1m :
                               totalPrice > 500 ? totalPrice * 0.05m : 0m;
                
                // Find top category in one pass
                var topCategory = orderItems
                    .GroupBy(item => item.Category)
                    .Select(g => new { Category = g.Key, Count = g.Sum(i => i.Quantity) })
                    .OrderByDescending(x => x.Count)
                    .FirstOrDefault()?.Category;
                
                // Create summary
                return new OrderSummary
                {
                    OrderId = order.OrderId,
                    CustomerId = order.CustomerId,
                    OrderDate = order.OrderDate,
                    ItemCount = itemCount,
                    TotalAmount = totalPrice,
                    Discount = discount,
                    NetAmount = totalPrice - discount,
                    TopCategory = topCategory
                };
            }).ToList();
        }
        
        /// <summary>
        /// Process orders in parallel for large datasets.
        /// Uses Parallel LINQ (PLINQ) for better performance with large data sets.
        /// </summary>
        public IReadOnlyList<OrderSummary> ProcessOrdersParallel(IEnumerable<Order> orders, int minOrdersForParallel = 1000)
        {
            if (orders == null)
            {
                return Array.Empty<OrderSummary>();
            }
            
            // Convert to array once to avoid multiple enumeration
            var orderArray = orders as Order[] ?? orders.ToArray();
            
            // Only use parallel processing for large datasets
            if (orderArray.Length < minOrdersForParallel)
            {
                return ProcessOrders(orderArray);
            }
            
            // Use PLINQ for parallel processing
            return orderArray.AsParallel()
                .Select(order =>
                {
                    // Calculate total price and item count
                    var orderItems = order.Items;
                    var totalPrice = orderItems.Sum(item => item.Price * item.Quantity);
                    var itemCount = orderItems.Sum(item => item.Quantity);
                    
                    // Calculate discount
                    var discount = totalPrice > 1000 ? totalPrice * 0.1m :
                                  totalPrice > 500 ? totalPrice * 0.05m : 0m;
                    
                    // Find top category
                    var topCategory = orderItems
                        .GroupBy(item => item.Category)
                        .Select(g => new { Category = g.Key, Count = g.Sum(i => i.Quantity) })
                        .OrderByDescending(x => x.Count)
                        .FirstOrDefault()?.Category;
                    
                    // Create summary
                    return new OrderSummary
                    {
                        OrderId = order.OrderId,
                        CustomerId = order.CustomerId,
                        OrderDate = order.OrderDate,
                        ItemCount = itemCount,
                        TotalAmount = totalPrice,
                        Discount = discount,
                        NetAmount = totalPrice - discount,
                        TopCategory = topCategory
                    };
                })
                .ToList();
        }
        
        /// <summary>
        /// Process orders asynchronously, optimized for memory efficiency.
        /// </summary>
        public async Task<IReadOnlyList<OrderSummary>> ProcessOrdersAsync(IAsyncEnumerable<Order> orders)
        {
            var results = new ConcurrentBag<OrderSummary>();
            
            await foreach (var order in orders)
            {
                // Process each order individually
                var orderItems = order.Items;
                
                // Efficiently calculate order metrics
                decimal totalPrice = 0;
                int itemCount = 0;
                var categoryCounts = new Dictionary<string, int>();
                
                foreach (var item in orderItems)
                {
                    totalPrice += item.Price * item.Quantity;
                    itemCount += item.Quantity;
                    
                    if (!categoryCounts.TryAdd(item.Category, item.Quantity))
                    {
                        categoryCounts[item.Category] += item.Quantity;
                    }
                }
                
                // Calculate discount
                var discount = totalPrice > 1000 ? totalPrice * 0.1m :
                              totalPrice > 500 ? totalPrice * 0.05m : 0m;
                
                // Find top category
                var topCategory = categoryCounts
                    .OrderByDescending(kvp => kvp.Value)
                    .FirstOrDefault().Key;
                
                // Create summary
                var summary = new OrderSummary
                {
                    OrderId = order.OrderId,
                    CustomerId = order.CustomerId,
                    OrderDate = order.OrderDate,
                    ItemCount = itemCount,
                    TotalAmount = totalPrice,
                    Discount = discount,
                    NetAmount = totalPrice - discount,
                    TopCategory = topCategory
                };
                
                results.Add(summary);
            }
            
            return results.ToList();
        }
    }

    /// <summary>
    /// Transaction class for the example
    /// </summary>
    public class Transaction
    {
        public string Id { get; set; }
        public DateTime Date { get; set; }
        public decimal Amount { get; set; }
        public string Merchant { get; set; }
        public string Category { get; set; }
    }
    
    /// <summary>
    /// Monthly statistics class
    /// </summary>
    public class MonthlyStatistics
    {
        public int Year { get; set; }
        public int Month { get; set; }
        public int TransactionCount { get; set; }
        public decimal TotalAmount { get; set; }
        public decimal AverageAmount { get; set; }
        public decimal LargestTransactionAmount { get; set; }
    }
    
    /// <summary>
    /// Order class for the example
    /// </summary>
    public class Order
    {
        public string OrderId { get; set; }
        public string CustomerId { get; set; }
        public DateTime OrderDate { get; set; }
        public List<OrderItem> Items { get; set; } = new List<OrderItem>();
    }
    
    /// <summary>
    /// Order item class
    /// </summary>
    public class OrderItem
    {
        public string ItemId { get; set; }
        public string Name { get; set; }
        public string Category { get; set; }
        public decimal Price { get; set; }
        public int Quantity { get; set; }
    }
    
    /// <summary>
    /// Order summary class
    /// </summary>
    public class OrderSummary
    {
        public string OrderId { get; set; }
        public string CustomerId { get; set; }
        public DateTime OrderDate { get; set; }
        public int ItemCount { get; set; }
        public decimal TotalAmount { get; set; }
        public decimal Discount { get; set; }
        public decimal NetAmount { get; set; }
        public string TopCategory { get; set; }
    }
} 

File: java-examples/scenario2-data-pipeline/stream-optimization/after/DataProcessingService.java
package com.enterprise.optimization.service;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;
import java.util.stream.Collectors;
import java.time.LocalDate;
import java.math.BigDecimal;
import java.util.Optional;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Function;
import java.util.function.Predicate;

/**
 * Service for data processing operations using optimized Java streams.
 * This implementation demonstrates best practices for efficient stream processing.
 * 
 * GitHub Copilot Agent Autonomous Mode:
 * To use with Copilot Agent, instruct it to: "Analyze this class for stream performance
 * bottlenecks and suggest improvements for handling large transaction volumes."
 */
public class DataProcessingService {

    /**
     * Process transaction data to calculate total amount by category.
     * Optimized implementation using a single stream operation with groupingBy collector.
     */
    public Map<String, BigDecimal> calculateTotalByCategory(List<Transaction> transactions) {
        if (transactions == null || transactions.isEmpty()) {
            return new HashMap<>();
        }
        
        // Single stream operation with collectors for better performance
        return transactions.stream()
            .filter(t -> t.getAmount().compareTo(BigDecimal.ZERO) > 0) // Filter valid transactions
            .collect(Collectors.groupingBy(
                Transaction::getCategory,
                Collectors.reducing(BigDecimal.ZERO, Transaction::getAmount, BigDecimal::add)
            ));
    }
    
    /**
     * Find transactions that meet specific criteria.
     * Optimized implementation using a single stream with multiple filters.
     */
    public List<Transaction> findTransactionsByCriteria(List<Transaction> transactions, 
                                                       String merchant, 
                                                       LocalDate startDate, 
                                                       LocalDate endDate, 
                                                       BigDecimal minAmount) {
        if (transactions == null || transactions.isEmpty()) {
            return new ArrayList<>();
        }
        
        // Using a single stream with all filters combined
        return transactions.stream()
            .filter(t -> t.getMerchant().equalsIgnoreCase(merchant))
            .filter(t -> !t.getDate().isBefore(startDate) && !t.getDate().isAfter(endDate))
            .filter(t -> t.getAmount().compareTo(minAmount) >= 0)
            .collect(Collectors.toList());
    }
    
    /**
     * Group transactions by month and calculate monthly statistics.
     * Optimized implementation with efficient collectors and reduced object creation.
     */
    public List<MonthlyStatistics> calculateMonthlyStatistics(List<Transaction> transactions) {
        if (transactions == null || transactions.isEmpty()) {
            return new ArrayList<>();
        }
        
        // Create a reusable YearMonth function to avoid creating duplicate objects
        Function<Transaction, YearMonth> yearMonthExtractor = 
            t -> new YearMonth(t.getDate().getYear(), t.getDate().getMonthValue());
        
        // Using advanced groupingBy collector with downstream collectors
        Map<YearMonth, List<Transaction>> transactionsByMonth = transactions.stream()
            .collect(Collectors.groupingBy(yearMonthExtractor));
        
        // Process each month's statistics in a more efficient manner
        return transactionsByMonth.entrySet().stream()
            .map(entry -> {
                YearMonth yearMonth = entry.getKey();
                List<Transaction> monthTransactions = entry.getValue();
                
                // Calculate all statistics in a single stream pass
                MonthlyStatistics stats = new MonthlyStatistics();
                stats.yearMonth = yearMonth;
                stats.transactionCount = monthTransactions.size();
                
                // Get total and find largest transaction in a single pass using custom collector
                MonthlyStats monthlyStats = monthTransactions.stream()
                    .collect(
                        // Initial result supplier
                        () -> new MonthlyStats(BigDecimal.ZERO, null),
                        // Accumulator
                        (result, transaction) -> {
                            // Add amount to total
                            result.total = result.total.add(transaction.getAmount());
                            
                            // Update max transaction if needed
                            if (result.max == null || 
                                transaction.getAmount().compareTo(result.max.getAmount()) > 0) {
                                result.max = transaction;
                            }
                        },
                        // Combiner (for parallel streams)
                        (result1, result2) -> {
                            result1.total = result1.total.add(result2.total);
                            if (result1.max == null ||
                                (result2.max != null && 
                                 result2.max.getAmount().compareTo(result1.max.getAmount()) > 0)) {
                                result1.max = result2.max;
                            }
                        }
                    );
                
                stats.totalAmount = monthlyStats.total;
                stats.largestTransactionAmount = 
                    monthlyStats.max != null ? monthlyStats.max.getAmount() : BigDecimal.ZERO;
                
                // Calculate average amount
                if (stats.transactionCount > 0) {
                    stats.averageAmount = stats.totalAmount
                        .divide(BigDecimal.valueOf(stats.transactionCount), 2, BigDecimal.ROUND_HALF_UP);
                } else {
                    stats.averageAmount = BigDecimal.ZERO;
                }
                
                return stats;
            })
            .collect(Collectors.toList());
    }
    
    /**
     * Process transactions in parallel for large datasets.
     * Uses parallel streams with appropriate safeguards.
     */
    public Map<String, BigDecimal> calculateTotalByCategoryParallel(List<Transaction> transactions) {
        if (transactions == null || transactions.isEmpty()) {
            return new HashMap<>();
        }
        
        // For large datasets, parallel streams can provide performance benefits
        return transactions.parallelStream()
            .filter(t -> t.getAmount().compareTo(BigDecimal.ZERO) > 0)
            .collect(Collectors.groupingByConcurrent(
                Transaction::getCategory,
                Collectors.reducing(BigDecimal.ZERO, Transaction::getAmount, BigDecimal::add)
            ));
    }
    
    /**
     * Filter unique transactions using a stateful predicate.
     * Uses a more efficient approach to remove duplicates.
     */
    public List<Transaction> filterUniqueTransactions(List<Transaction> transactions) {
        if (transactions == null || transactions.isEmpty()) {
            return new ArrayList<>();
        }
        
        // Create a stateful predicate to track seen IDs
        Predicate<Transaction> distinctById = 
            distinctByKey(Transaction::getId);
        
        // Filter transactions keeping only the first occurrence of each ID
        return transactions.stream()
            .filter(distinctById)
            .collect(Collectors.toList());
    }
    
    /**
     * Helper method to create a stateful predicate for distinct filtering by key.
     */
    private static <T> Predicate<T> distinctByKey(Function<? super T, ?> keyExtractor) {
        Map<Object, Boolean> seen = new ConcurrentHashMap<>();
        return t -> seen.putIfAbsent(keyExtractor.apply(t), Boolean.TRUE) == null;
    }
    
    /**
     * Helper class for collecting monthly statistics in a single pass
     */
    private static class MonthlyStats {
        BigDecimal total;
        Transaction max;
        
        MonthlyStats(BigDecimal total, Transaction max) {
            this.total = total;
            this.max = max;
        }
    }
    
    /**
     * Transaction class for the example
     */
    public static class Transaction {
        private String id;
        private LocalDate date;
        private BigDecimal amount;
        private String merchant;
        private String category;
        
        // Getters and setters
        public String getId() { return id; }
        public void setId(String id) { this.id = id; }
        
        public LocalDate getDate() { return date; }
        public void setDate(LocalDate date) { this.date = date; }
        
        public BigDecimal getAmount() { return amount; }
        public void setAmount(BigDecimal amount) { this.amount = amount; }
        
        public String getMerchant() { return merchant; }
        public void setMerchant(String merchant) { this.merchant = merchant; }
        
        public String getCategory() { return category; }
        public void setCategory(String category) { this.category = category; }
    }
    
    /**
     * Year-Month helper class
     */
    public static class YearMonth {
        public int year;
        public int month;
        
        public YearMonth(int year, int month) {
            this.year = year;
            this.month = month;
        }
        
        @Override
        public boolean equals(Object obj) {
            if (!(obj instanceof YearMonth)) return false;
            YearMonth other = (YearMonth) obj;
            return this.year == other.year && this.month == other.month;
        }
        
        @Override
        public int hashCode() {
            return 31 * year + month;
        }
    }
    
    /**
     * Monthly statistics class
     */
    public static class MonthlyStatistics {
        public YearMonth yearMonth;
        public int transactionCount;
        public BigDecimal totalAmount;
        public BigDecimal averageAmount;
        public BigDecimal largestTransactionAmount;
    }
} 